File: lib/services/reasoning_effort.dart
// Copyright (C) 2026 Front Porch AI
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// App-facing thinking strength (low / medium / high) and how it maps onto
// whatever ladder a remote model actually accepts. Shared by the remote
// payload builder and the settings UI so the labels users see match the
// wire behaviour.

/// App surface values — what the user picks in Settings / chat settings.
const List<String> kAppReasoningEfforts = ['low', 'medium', 'high'];

/// Full ladder any provider has used (for nearest-match math).
const Map<String, int> kReasoningEffortRank = {
  'none': 0,
  'minimal': 1,
  'low': 2,
  'medium': 3,
  'high': 4,
  'max': 5,
};

/// Short display title for a strength id.
String reasoningEffortTitle(String id) => switch (id) {
      'none' => 'Off',
      'minimal' => 'Minimal',
      'low' => 'Low',
      'medium' => 'Medium',
      'high' => 'High',
      'max' => 'Max',
      _ => id,
    };

/// One-line blurb under a strength chip.
String reasoningEffortBlurb(String id) => switch (id) {
      'low' => 'Light think — faster, cheaper',
      'medium' => 'Balanced — default',
      'high' => 'Deep think — slower, richer',
      'none' => 'No thinking tokens',
      'max' => 'Full thinking budget',
      'minimal' => 'Bare-minimum thinking',
      _ => '',
    };

/// Models whose provider taught us a supported set (process lifetime).
/// Written by [OpenRouterService] when a 400 lists supported values.
final Map<String, Set<String>> kLearnedReasoningEffortsByModel =
    <String, Set<String>>{};

/// Nano-style `:thinking` suffix often only accepts none / high / max.
const Set<String> kThinkingSuffixEffortHint = {'none', 'high', 'max'};

/// Hint (not a hard table): ids with `:thinking` commonly lack low/medium.
Set<String>? reasoningEffortHintForModel(String model) {
  if (model.toLowerCase().contains(':thinking')) {
    return kThinkingSuffixEffortHint;
  }
  return null;
}

/// Supported set for [model]: learned 400 > `:thinking` hint > null (verbatim).
Set<String>? reasoningEffortSupportedFor(String model) {
  if (model.isEmpty) return null;
  return kLearnedReasoningEffortsByModel[model] ??
      reasoningEffortHintForModel(model);
}

/// Closest supported stand-in. Keeps thinking ON — never maps to `none` when
/// any thinking tier exists.
String nearestReasoningEffort(String requested, Set<String> supported) {
  if (supported.contains(requested)) return requested;
  final req = kReasoningEffortRank[requested] ?? kReasoningEffortRank['medium']!;
  final thinking = supported.where((s) => s != 'none');
  final pool = (thinking.isNotEmpty ? thinking : supported)
      .where(kReasoningEffortRank.containsKey);
  if (pool.isEmpty) return requested;
  return pool.reduce((a, b) {
    final da = (kReasoningEffortRank[a]! - req).abs();
    final db = (kReasoningEffortRank[b]! - req).abs();
    if (da != db) return da < db ? a : b;
    return kReasoningEffortRank[a]! < kReasoningEffortRank[b]! ? a : b;
  });
}

/// Value that will go on the wire for [model] given the user's [requested].
String wireReasoningEffort(String model, String requested) {
  final supported = reasoningEffortSupportedFor(model);
  if (supported == null) return requested;
  return nearestReasoningEffort(requested, supported);
}

/// True when the app pick differs from what the model will receive.
bool reasoningEffortIsRemapped(String model, String requested) {
  if (model.isEmpty || requested.isEmpty) return false;
  return wireReasoningEffort(model, requested) != requested;
}

/// User-facing caption: what this strength becomes on the current model.
///
/// Empty when there is no model context or no remap.
String reasoningEffortMappingCaption(String model, String requested) {
  if (model.isEmpty || requested.isEmpty) return '';
  final supported = reasoningEffortSupportedFor(model);
  if (supported == null) {
    return 'Sent as ${reasoningEffortTitle(requested).toLowerCase()}. '
        'If this model only accepts other levels, Front Porch maps to the '
        'closest one automatically.';
  }
  final wire = nearestReasoningEffort(requested, supported);
  final allowed = supported
      .where((s) => s != 'none')
      .map(reasoningEffortTitle)
      .join(' · ');
  if (wire == requested) {
    return 'This model accepts: $allowed. '
        'Your pick matches — sent as ${reasoningEffortTitle(wire).toLowerCase()}.';
  }
  return 'This model accepts: $allowed. '
      '${reasoningEffortTitle(requested)} → sent as '
      '${reasoningEffortTitle(wire).toLowerCase()}.';
}

/// Parse "Supported values are: none, high, max" from a provider error body.
Set<String>? supportedReasoningEffortsFromError(String msg) {
  final m = msg.toLowerCase();
  if (!m.contains('reasoning.effort') && !m.contains('reasoning effort')) {
    return null;
  }
  final listing = m.indexOf('supported values');
  if (listing < 0) return null;
  final colon = m.indexOf(':', listing);
  if (colon < 0) return null;
  final values = RegExp(r'[a-z]+')
      .allMatches(m.substring(colon + 1))
      .map((x) => x.group(0)!)
      .where(kReasoningEffortRank.containsKey)
      .toSet();
  return values.isEmpty ? null : values;
}

/// Remember a provider's supported set for [model] (process lifetime).
void rememberReasoningEffortsForModel(String model, Set<String> supported) {
  if (model.isEmpty || supported.isEmpty) return;
  kLearnedReasoningEffortsByModel[model] = Set<String>.from(supported);
}

File: web_ui/src/utils/reasoningEffort.ts
/** App-facing thinking strength (mirrors lib/services/reasoning_effort.dart). */

export const APP_REASONING_EFFORTS = ['low', 'medium', 'high'] as const
export type AppReasoningEffort = (typeof APP_REASONING_EFFORTS)[number]

const RANK: Record<string, number> = {
  none: 0,
  minimal: 1,
  low: 2,
  medium: 3,
  high: 4,
  max: 5,
}

const THINKING_HINT = new Set(['none', 'high', 'max'])

export function reasoningEffortTitle(id: string): string {
  switch (id) {
    case 'none': return 'Off'
    case 'minimal': return 'Minimal'
    case 'low': return 'Low'
    case 'medium': return 'Medium'
    case 'high': return 'High'
    case 'max': return 'Max'
    default: return id
  }
}

export function reasoningEffortBlurb(id: string): string {
  switch (id) {
    case 'low': return 'Light think — faster, cheaper'
    case 'medium': return 'Balanced — default'
    case 'high': return 'Deep think — slower, richer'
    default: return ''
  }
}

function supportedFor(model: string): Set<string> | null {
  if (!model) return null
  if (model.toLowerCase().includes(':thinking')) return THINKING_HINT
  return null
}

function nearest(requested: string, supported: Set<string>): string {
  if (supported.has(requested)) return requested
  const req = RANK[requested] ?? RANK.medium
  const thinking = [...supported].filter((s) => s !== 'none')
  const pool = (thinking.length ? thinking : [...supported]).filter((s) => s in RANK)
  if (!pool.length) return requested
  return pool.reduce((a, b) => {
    const da = Math.abs(RANK[a]! - req)
    const db = Math.abs(RANK[b]! - req)
    if (da !== db) return da < db ? a : b
    return RANK[a]! < RANK[b]! ? a : b
  })
}

export function wireReasoningEffort(model: string, requested: string): string {
  const supported = supportedFor(model)
  if (!supported) return requested
  return nearest(requested, supported)
}

export function reasoningEffortMappingCaption(model: string, requested: string): string {
  if (!requested) return ''
  const supported = supportedFor(model)
  if (!supported) {
    return 'Sent as-is. If this model only accepts other levels, Front Porch maps to the closest one automatically.'
  }
  const wire = nearest(requested, supported)
  const allowed = [...supported]
    .filter((s) => s !== 'none')
    .map(reasoningEffortTitle)
    .join(' · ')
  if (wire === requested) {
    return `This model accepts: ${allowed}. Your pick matches — sent as ${reasoningEffortTitle(wire).toLowerCase()}.`
  }
  return `This model accepts: ${allowed}. ${reasoningEffortTitle(requested)} → sent as ${reasoningEffortTitle(wire).toLowerCase()}.`
}
